rfcs/20190411-string-unification.md

```diff
@@ -0,0 +1,302 @@
+or `PreallocType` the string is converted to `LargeType` string.  The original
+string type is copied as a prefix to the `LargeType` so that it can be reverted
+when the capacity falls below the original on a subsequent assignment.  This
+feature will allow for lvalue assignment of OffsetType types, which, as a
+corollary, will allow for the assignment of TFLite string tensors.
+
+### Comparison of SSO implementations in gcc, llvm, and MSVC
+
+|                                           | Size in Bytes | Small String Capacity
+|------------------------------------------ | --            | --
+|tensorflow::string                         | 16            | 15
+|tensorflow::string w/ capacity (see below) | 24            | 23
+|GCC                                        | 32            | 15
+|MSVC                                       | 32            | 15
+|LLVM                                       | 24            | 22
+
+### Capacity member variable
+
+Currently, TFLite has an ~8 byte overhead per string entry, which is used to
+describe the offset.  The tstring class described above has an overhead of 16
+bytes, and does not include a capacity field normally found in SSO
+implementations for LargeType strings.  This comes with downsides.  Without a
+capacity field LargeType strings are forced to always call realloc on
+assignment.  To reduce the potential number of calls to realloc, we can add a
+capacity field, at the cost of increasing the per string overhead to 24 bytes.
+This would put as at parity with LLVM strings, but would result in a 3x
+overhead compared to current TFLite strings.
+
+### Updates to C API
+
+Currently, downstream language bindings to
+[Java](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/java/src/main/native/tensor_jni.cc?#L270),
+[golang](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/go/tensor.go?#L395),
+etc, are expected to
+modify a raw buffer in order to construct and pass a string tensor.  We propose
+new functions to abstract the creation of C string tensors, and a new flag in
+`TF_Tensor` which tracks the method with which a C string tensor was created.
+Using the new methods will effectively create C++ string tensors underneath,
+and, when passed back and forth, will mitigate the conversion of `C TF_STRING`
+tensors to C++ `DT_STRING` tensors and vice-versa via
+[`TF_TensorToTensor`](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/c/c_api.cc?#L417)
+and
+[`TF_TensorFromTensor`](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/c/c_api.cc?#L489).
+
+### Updates to TFLite
+
+Since TFLite provides generators and accessors for TFLite string tensors, the
+requisite changes needed to have TFLite conform to the `OffsetType` defined above
+is on the order of a ~20 line CL.  The more difficult task is to find all uses
+of string tensors (internally, and potentially) in flat files, and coordinate
```

Comment:
Echoing the above point, we cannot rely on updating all models. We'll have to decide on an approach which gracefully handles the existing layout for models that exist in the wild.

Reply:
I included your solution of adding a new TFLite enum type for tstring, which should gracefully handle existing existing layouts.